apollo-router-core/src/error.rs

```diff
@@ -132,6 +129,54 @@ pub struct Error {
     pub extensions: Object,
 }
 
+/// temporary structure to help deserializing errors
+#[derive(Deserialize)]
+struct ErrorMeta {
+    message: String,
+    locations: Vec<Location>,
+    path: Option<Path>,
+}
+
+impl Error {
+    pub fn from_value(service_name: &str, mut value: Value) -> Result<Error, FetchError> {
+        // get the extensions object manually because from_value would end up reallocating all the strings
+        let extensions = if let Value::Object(object) = &mut value {
+            match object.remove("extensions") {
+                Some(Value::Object(o)) => o,
+                _ => {
+                    return Err(FetchError::SubrequestMalformedResponse {
+                        service: service_name.to_string(),
+                        reason: "missing extensions field".to_string(),
+                    })
+                }
+            }
+        } else {
+            return Err(FetchError::SubrequestMalformedResponse {
+                service: service_name.to_string(),
+                reason: "expected a JSON object".to_string(),
+            });
+        };
```

Comment:
:yarn: I personally like the code to do one thing when possible and optimal. In this change I separated the check that makes sure the value received is an object from the extraction of the key extensions. I think it's clearer and easier to understand.

```suggestion
        let object = if let Value::Object(object) = value {
            object
        } else {
            return Err(FetchError::SubrequestMalformedResponse {
                service: service_name.to_string(),
                reason: "expected a JSON object".to_string(),
            });
        };

        // get the extensions object manually because from_value would end up reallocating all the strings
        let extensions = match object.remove("extensions") {
            Some(Value::Object(o)) => o,
            _ => {
                return Err(FetchError::SubrequestMalformedResponse {
                    service: service_name.to_string(),
                    reason: "missing extensions field (or not an object)".to_string(),
                })
            }
        };
```

Reply:
it would indeed be clearer, but here we must not consume `value`, only extract the `extensions` field, because `value` is used here: https://github.com/apollographql/router/pull/284/files#diff-3fee2c973320bfc619c5c548a725cfffbc3874e50bbd76351eca927d8f89ee6eR164

There's probably a bettter way to write it even with that constraint

Reply:
I guess you can just wrap it again into `Value::Object(...)` :thinking: not sure if it would have any performance impact. I don't mind keeping it that way for now

Reply:
oh, right, that could work :grinning: